File: python3/vaccine.py
from collections import deque

def solve(f):
    RNAList = [list(line) for line in f]
    testCases = len(RNAList)

    for i in range(1, testCases):
        #print(RNAList[i])

        queue = deque()
        I = len(RNAList[i]) - 2

        A1 = False
        U1 = False
        G1 = False
        C1 = False
        if RNAList[i][I] == 'A' : A1 = True
        if RNAList[i][I] == 'U' : U1 = True
        if RNAList[i][I] == 'G' : G1 = True
        if RNAList[i][I] == 'C' : C1 = True

        queue.append((False, I - 1 ,RNAList[i] , RNAList[i][I], RNAList[i][I], 'p','p', 'Z', A1, U1, G1, C1))
        
        while queue:
            comp, I, RNA, head, tail, result, prev1, prev2, A, U, G, C = queue.popleft()

            #print(comp, I, RNA, head, tail, result, prev1, prev2, A, U, G, C)

            if I == -1:
                break
            
            if comp:
                if RNA[I] == 'A': cur = 'U'
                if RNA[I] == 'U': cur = 'A'
                if RNA[I] == 'G': cur = 'C'
                if RNA[I] == 'C': cur = 'G'
                #print("---------------> new cur = ", cur)
            else : cur = RNA[I]

            if (prev1 != 'c' and not((prev1 == 'r' and prev2 == 'c') or (prev1 == 'c' and prev2 == 'r')) and prev1 != 'r'):
                queue.append((not comp, I, RNA, head, tail, result + 'c', 'c', prev1, A, U, G, C))
                #print(2)

            if head == cur or ((cur == 'A' and A == False) or (cur == 'U' and U == False) or (cur == 'G' and G == False) or (cur == 'C' and C == False)):
                A1 = A
                U1 = U
                G1 = G
                C1 = C
                if cur == 'A' : A1 = True
                if cur == 'U' : U1 = True
                if cur == 'G' : G1 = True
                if cur == 'C' : C1 = True
                if tail == 'Z' : tail = cur
                queue.append((comp, I-1, RNA, cur, tail, result + 'p', 'p', prev1, A1, U1, G1, C1))
                #print(1)


            if (prev1 != 'r' and not ((prev1 == 'r' and prev2 == 'c') or (prev1 == 'c' and prev2 == 'r'))):
                queue.append((comp, I, RNA, tail, head, result + 'r', 'r', prev1, A, U, G, C))
                #print(3)

            

        print(result)



if __name__ == '__main__':
    import sys
    with open(sys.argv[1], 'rt') as f:
        solve(f)

File: prolog/vaccine.pl
empty(queue([],[])).
add(A,queue(X,Y),queue(X,[A|Y])).
remove(queue([],Z),A,queue(X,[])) :- reverse(Z,[A|X]).
remove(queue([A|T],X),A,queue(T,X)).

read_input(File, N, C) :-
    open(File, read, Stream),
    read_line(Stream, N),
    read_rest(Stream, N, C).

read_line(Stream, L) :-
    read_line_to_codes(Stream, Line),
    atom_codes(Atom, Line),
    atomic_list_concat(Atoms, ' ', Atom),
    maplist(atom_number, Atoms, L).

read_rest(_, [0], []):- !.
read_rest(Stream, [N], C):-
    read_string(Stream, "\n", "\r", _, String),
    string_chars(String, List),
    N1 is N - 1,
    read_rest(Stream, [N1], C1),
    append([List], C1, C).

bfs((_, [], _, _, RET, _, _, _, _, _, _),_, RET).
bfs((Comp, [H|T], Head, Tail, Res, Prev1, Prev2, A, U, G, C),Queue, RET):-
    (Comp == 1 -> 
        ((H == 'A' -> Cur = 'U');
        (H == 'U' -> Cur = 'A');
        (H == 'G' -> Cur = 'C');
        (H == 'C' -> Cur = 'G'));
        Cur = H),
            
    (Prev1 == 'p' ->
        NC is Comp * -1,
        add((NC, [H|T], Head, Tail, ['c'|Res], 'c', Prev1, A, U, G, C), Queue, Q1)
    ;Q1 = Queue),
    
    ((Head == Cur ; ((Cur == 'A' , A == 0) ; (Cur == 'U' , U == 0) ; (Cur == 'G' , G == 0) ; (Cur == 'C' , C == 0))) -> 
        ((Cur == 'A' -> A1 is 1, U1 is U, G1 is G, C1 is C);
        (Cur == 'U' -> A1 is A, U1 is 1, G1 is G, C1 is C);
        (Cur == 'G' -> A1 is A, U1 is U, G1 is 1, C1 is C);
        (Cur == 'C' -> A1 is A, U1 is U, G1 is G, C1 is 1)),
        add((Comp, T, Cur, Tail, ['p'|Res], 'p', Prev1, A1,U1, G1, C1), Q1, Q2)
    ;Q2 = Q1),

    (Prev1 \== 'r', \+ ((Prev1 == 'r', Prev2 == 'c') ; (Prev1 == 'c', Prev2 == 'r')) ->
        add((Comp, [H|T], Tail, Head, ['r'|Res], 'r', Prev1, A, U, G, C), Q2, Q3)
    ;Q3 = Q2),

    remove(Q3, Next, Q4),
    bfs(Next, Q4, RET).

solve([], []).
solve([H|T], A):-
    reverse(H, [First|Hrev]),
    ((First == 'A' -> A1 is 1, U1 is 0, G1 is 0, C1 is 0);
    (First == 'U' -> A1 is 0, U1 is 1, G1 is 0, C1 is 0);
    (First == 'G' -> A1 is 0, U1 is 0, G1 is 1, C1 is 0);
    (First == 'C' -> A1 is 0, U1 is 0, G1 is 0, C1 is 1)),
    empty(Queue),
    bfs((-1, Hrev, First, First, ['p'], 'p', 'Z', A1, U1, G1, C1), Queue,RET),
    solve(T, NewA),
    reverse(RET, RETT), 
    string_chars(S, RETT),
    
    append([S], NewA, A).   

vaccine(File, A):-
    read_input(File, _, C),
    solve(C, A).
